src/query/storages/common/index/src/filters/xor8/bloom_filter.rs

```diff
@@ -0,0 +1,288 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::max;
+use std::collections::HashSet;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::mem;
+
+use anyerror::AnyError;
+use databend_common_exception::ErrorCode;
+use databend_common_expression::types::DataType;
+use databend_common_functions::scalars::CityHasher64;
+
+use crate::filters::Filter;
+use crate::filters::FilterBuilder;
+use crate::Index;
+
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+type UnderType = u64;
+
+pub struct BloomBuilder {
+    bloom_size: usize,
+    seed: u64,
+    inner: HashSet<u64>,
+}
+
+impl BloomBuilder {
+    pub fn create(bloom_size: usize, seed: u64) -> Self {
+        Self {
+            bloom_size,
+            seed,
+            inner: Default::default(),
+        }
+    }
+}
+
+#[derive(thiserror::Error, serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+#[error("{msg}")]
+pub struct BloomCodecError {
+    msg: String,
+}
+
+#[derive(thiserror::Error, serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+#[error("fail to build bloom filter; cause: {cause}")]
+pub struct BloomBuildingError {
+    #[source]
+    cause: AnyError,
+}
+
+impl FilterBuilder for BloomBuilder {
+    type Filter = BloomFilter;
+    type Error = BloomBuildingError;
+
+    fn add_key<K: Hash>(&mut self, key: &K) {
+        let mut hasher64 = CityHasher64::with_seed(self.seed);
+        key.hash(&mut hasher64);
+        self.inner.insert(hasher64.finish());
+    }
+
+    fn add_keys<K: Hash>(&mut self, keys: &[K]) {
+        for key in keys {
+            self.add_key::<K>(key);
+        }
+    }
+
+    fn add_digests<'i, I: IntoIterator<Item = &'i u64>>(&mut self, digests: I) {
+        self.inner.extend(digests);
+    }
+
+    fn build(&mut self) -> Result<Self::Filter, Self::Error> {
+        let item_count = self.inner.len();
+
+        let mut filter = BloomFilter::with_item_count(self.bloom_size, item_count, self.seed);
+        for hash in mem::take(&mut self.inner) {
+            filter.add(hash);
+        }
+        Ok(filter)
+    }
+}
+
+#[derive(Clone)]
+pub struct BloomFilter {
+    size: usize,
```

Comment:
use `u64` instead of `usize`, because the size of `usize` is related to the platform, it may cause error during deserialization